### PR TITLE
chore(compass-crud): Unify interface between json editor and document editor views

### PR DESCRIPTION
**Pull request by gribnoysup**
Another round of clean-up in preparation for extraction of the document list views: this time I'm normalizing interfaces between json editor and and document editor components so that when we extract them we don't need to keep duplicated logic for different view types with slightly different interfaces.

As a side effect this also fixed some issues with the JSON editor:

- Update status is now bound to a single document by using document events (similar to document list view components) instead of being shared between all the cards (which lead to confusing behavior if multiple cards were in the editing state)
- Editing timeseries documents was allowed in JSON view because isTimeSeries was never taken into account for the JSON list view, now we are handling it the same it is handled for document list view

### CODE REVIEW

File: packages/compass-crud/src/components/json-editor.jsx [hunk 1]
```diff
@@ -94,87 +65,104 @@ class EditableJson extends React.Component {
...
-      this.editor.getSession().foldAll(2);
-    }
-    if (this.state.editing && this.props.updateError) {
-      this.handleUpdateError();
-    } else if (this.state.deleting && this.props.updateSuccess) {
-      this.handleRemoveSuccess();
-    } else if (this.state.editing && this.props.updateSuccess) {
-      this.handleUpdateSuccess();
-    } else if (this.props.doc !== prevProps.doc) {
+  componentDidUpdate(prevProps, prevState) {
```
Comment (Anemy) [inline packages/compass-crud/src/components/json-editor.jsx:89]: Looks like there are two `componentDidUpdate` methods now, should we consolidate them?

Comment (gribnoysup) [reply packages/compass-crud/src/components/json-editor.jsx:89]: Oh lol, I swear this react components methods sort order eslint rule makes me do the silliest things. Thanks for spotting, these should indeed just be one now

File: packages/compass-crud/src/components/document-footer.jsx [hunk 2]
```diff
@@ -247,9 +260,12 @@ DocumentFooter.displayName = 'DocumentFooter';
...
   doc: PropTypes.object.isRequired,
   replaceDocument: PropTypes.func.isRequired,
-  updateDocument: PropTypes.func.isRequired,
+  updateDocument: PropTypes.func,
   cancelHandler: PropTypes.func,
-  api: PropTypes.any
+  api: PropTypes.any,
+  editing: PropTypes.bool,
+  modified: PropTypes.bool,
+  containsErrors: PropTypes.bool
```
Comment (gribnoysup) [inline packages/compass-crud/src/components/document-footer.jsx:268]: This sorta allows for controlled / uncontrolled behavior for this footer component which is still probably messier than it should be, but I'll be tackling footer extraction separately so feels like this implementation is okay for now (but please tell me if you think that it's not)

File: packages/hadron-document/index.d.ts [hunk 3]
```diff
@@ -112,10 +112,12 @@ declare const HadronDocumentEvents: {
 
 declare class HadronDocument extends EventEmitter {
   constructor(doc: unknown, cloned: boolean);
+  uuid: string;
   type: 'Document';
   currentType: 'Document';
   elements: LinkedList<HadronElement>;
   cancel(): void;
+  apply(doc: unknown): void;
```
Comment (addaleax) [inline packages/hadron-document/index.d.ts:120]: Maybe `insertBeginning()` should be added here as well?

Comment (gribnoysup) [reply packages/hadron-document/index.d.ts:120]: Totally, thanks! Added in https://github.com/mongodb-js/compass/pull/2929/commits/8586328d38076aaa5f6b9508d3395c6a507bb582

**Review by lerouxb — APPROVED** [review]

**Pull request merged**